src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +871,135 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+            co_await apply_retention();
+            co_await garbage_collect();
+        }
+    } catch (std::exception& e) {
+        vlog(_rtclog.warn, "Error occured during housekeeping", e.what());
+    }
+}
+
+ss::future<> ntp_archiver::apply_retention() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    auto retention_calculator = retention_calculator::factory(
+      manifest(), _partition->get_ntp_config());
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        vlog(
+          _rtclog.debug,
+          "{} Advancing start offset to {} satisfy retention policy",
+          retention_calculator->strategy_name(),
+          *next_start_offset);
+
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
+          *next_start_offset, deadline, _as);
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.warn,
+              "Failed to update archival metadata STM start offest according "
+              "to retention policy: {}",
+              error);
+        }
+    } else {
+        vlog(
+          _rtclog.debug,
+          "{} Retention policies are already met.",
+          retention_calculator->strategy_name());
+    }
+}
+
+// Garbage collection can be improved as follows:
+// * issue #6843: delete via DeleteObjects S3 api instead of deleting individual
+// segments
+// * issue #6844: flush _replaced and advance the start offset even if the
+// deletions fail if the backlog breaches a certain limit. This can lead to
+// segments begin orphaned, but it's preferable to unbounded backlog growth.
+ss::future<> ntp_archiver::garbage_collect() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    std::vector<remote_segment_path> to_remove;
+    auto current_start_offset = manifest().get_start_offset();
+
+    for (const auto& [key, value] : manifest()) {
+        if (value.committed_offset < current_start_offset) {
+            to_remove.emplace_back(
+              manifest().generate_remote_segment_path(_ntp, key, value));
+        }
+    }
+
+    auto replaced = manifest().replaced_segments();
+    for (const auto& meta : replaced) {
+        const auto segment_name = cloud_storage::generate_local_segment_name(
+          meta.base_offset, meta.segment_term);
+        to_remove.emplace_back(cloud_storage::generate_remote_segment_path(
+          _ntp, meta.ntp_revision, segment_name, meta.archiver_term));
+    }
+
+    std::atomic<size_t> successful_deletes{0};
+    ssx::semaphore limiter{_concurrency, "archival/delete"};
+    for (const auto& path : to_remove) {
+        auto units = co_await ss::get_units(limiter, 1);
```

Comment:
I don't think that this achieves the desired goal. The loop will still be sequential. I think that you might want to use `ss::coroutine::all` or `ss::max_concurrent_for_each` here.

Reply:
Yep. You're right. Doesn't actually start futures in parallel. Thanks for pointing it out.